Rewrite synology camera by using py-synology package

snjoetw opened the pull request:
## Description:
This is to rewrite Synology camera component so that Synology APIs are hidden from HA device by using py-synology package. I've also added more features to this component:
- show correct camera state based on is_recording/is_streaming flag
- show correct motion detection status in device attribute
- added support to enable / disable motion detection
- periodically polls from Synology APIs to get latest camera states

## Checklist:

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [ ] New files were added to `.coveragerc`.


[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

snjoetw committed:
- Rewrite synology camera by intruducing Api and SurveillanceStation classes to get cameras, motion settings, enable/disable motion detection, etc ...
- Synology camera now shows correct state based on is_recording and is_streaming flag. Also it now supports enable / disable motion detection and show the correct motion detection status
- Newly added Api and SurveillanceStation classes will be moved to a lib but it's here just for review

snjoetw committed:
- Updated how payload are merged with kwargs so it works with python <3.5

snjoetw committed:
- Fixed class name conflict

snjoetw committed:
- Addressed flake8 error

snjoetw committed:
- Addressed pylint error

MartinHjelmare commented:
Just some general advice on commit messages and PR names:

- The first line of the commit message is the header and should be within 50 characters. Don't end it with a period. Write the header in imperative mode: "Add new feature X". Don't start all commit messages with a dash.
- Separate the header from the body with a blank line.
- The body of the commit message should be within 72 characters. The body you right as you want. Markdown is nice.
- Github will automatically use the header of your commit message as the name of the PR, if you only submit one commit when creating the PR. If multiple commits are present at the time the branch name will be used instead.
- Don't start the PR name with a dash. Write the name in imperative mode like the commit message header.

MartinHjelmare commented on homeassistant/components/camera/synology.py:
```diff
@@ -203,48 +93,348 @@ def camera_image(self):
+MOTION_DETECTION_SOURCE_BY_SURVEILLANCE = 1
+
+
+class Api:
```
All of the classes below should probably be broken out into a separate python package that is published on pypi.

snjoetw replied:
All those classes are in its own package now: https://github.com/snjoetw/py-synology

MartinHjelmare reviewed (changes requested):
Break the new classes out into a package that you publish on pypi.

Edit: Like you already pointed in the description above. Sorry for not reading everything first.

MartinHjelmare commented on homeassistant/components/camera/synology.py:
```diff
@@ -203,48 +93,348 @@ def camera_image(self):
+    },
+}
+
+API_NAMES = map(lambda api: api['name'], BASE_API_INFO.values())
```
Use a list comprehension instead.

snjoetw replied:
Sorry I'm new to python world so hopefully the change I made is what you meant: 
https://github.com/snjoetw/py-synology/blob/master/synology/api.py#L25

MartinHjelmare reviewed (commented):
The api looks really good after a quick look. @pvizeli might have better input on the async and camera specific parts.

snjoetw committed:
- Moved synology API related code to py-synology lib
- Added py-synology==0.1.1 requirement
- Removed hass from SynologyCamera constructor

snjoetw committed:
Merge branch 'dev' of https://github.com/home-assistant/home-assistant into syno_lib

snjoetw committed:
- Updated requirements_all.txt

snjoetw committed:
- renamed variable back to original

snjoetw commented:
Synology API related codes are moved to https://github.com/snjoetw/py-synology

I've added py-synology as the requirement for this component

snjoetw commented on homeassistant/components/camera/synology.py:
```diff
@@ -62,138 +44,46 @@ def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
-    def __init__(self, hass, websession, config, camera_id,
-                 camera_name, snapshot_path, streaming_path, camera_path,
-                 auth_path, timeout):
+    def __init__(self, hass, websession, surveillance, camera_id):
```
This is removed now

pvizeli commented on homeassistant/components/camera/synology.py:
```diff
@@ -203,48 +92,48 @@ def camera_image(self):
-        image = yield from response.read()
-
-        return image
+        return self._surveillance.get_camera_image(self._camera_id)
```
is that async?

snjoetw replied:
No this is synchronous call using requests.get()

pvizeli replied:
Please call this in sync context. Remove all async stuff from this function decleration

snjoetw replied:
Updated

pvizeli commented on homeassistant/components/camera/synology.py:
```diff
@@ -203,48 +92,48 @@ def camera_image(self):
+        return True
+
+    @asyncio.coroutine
+    def async_update(self):
```
It is not async

snjoetw replied:
Need your clarification on this

I see there's update() and async_update(), but I couldn't find where update() is called.
The only thing I found is async_update() is called in camera/__init__.py when polling is enabled:

```
    @asyncio.coroutine
    def async_handle_camera_service(service):
        """Handle calls to the camera services."""
        target_cameras = component.async_extract_from_service(service)

        for camera in target_cameras:
            if service.service == SERVICE_EN_MOTION:
                yield from camera.async_enable_motion_detection()
            elif service.service == SERVICE_DISEN_MOTION:
                yield from camera.async_disable_motion_detection()

        update_tasks = []
        for camera in target_cameras:
            if not camera.should_poll:
                continue

            update_coro = hass.async_add_job(
                camera.async_update_ha_state(True))
            if hasattr(camera, 'async_update'):
                update_tasks.append(update_coro)
            else:
                yield from update_coro

        if update_tasks:
            yield from asyncio.wait(update_tasks, loop=hass.loop)
```

I see there's a comment in camera.nest that the update doesn't seem to be called:
https://github.com/home-assistant/home-assistant/blob/2df433eb0a11752a8cb7b5bedbe87296539c801f/homeassistant/components/camera/nest.py#L74
I debugged it a bit and can confirm that I never seen update() being called when I left nest cam running for more than a day.

snjoetw replied:
I tried again with update() and looks like I was wrong, it did get called. So I've moved code in async_update() to update() now.

snjoetw replied:
But I still couldn't figure out where update() gets called in the code, would you point me where it is?

pvizeli replied:
helpers/entity.py

snjoetw replied:
Got it, thanks

Just for my reference: https://github.com/home-assistant/home-assistant/blob/2df433eb0a11752a8cb7b5bedbe87296539c801f/homeassistant/helpers/entity.py#L215

snjoetw committed:
- Sync call to retrieve camera image should be done in camera_image() instead

snjoetw committed:
- Sync call to update camera info should be done in update() instead

snjoetw committed:
- Removed unused import

pvizeli commented on homeassistant/components/camera/synology.py:
```diff
@@ -62,189 +43,89 @@ def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
+    def update(self):
+        """Update the status of the camera."""
+        self._surveillance.update()
+        self._camera = self._surveillance.get_camera(self._camera.camera_id)
```
Why that need do every X seconds?

snjoetw replied:
It's to fetch is_recording / is_enable value so that we show the correct camera state in the UI

pvizeli reviewed (approved):
Only a question is open, after that we can merge it
